app/models/miq_server/worker_management/monitor/kill.rb

```diff
@@ -5,13 +5,29 @@ def kill_all_workers
     return unless self.is_local?
 
     killed_workers = []
-    miq_workers.each do |w|
-      if MiqWorker::STATUSES_CURRENT_OR_STARTING.include?(w.status)
+
+    # We try but forget to write migrations to delete orphaned worker class rows, causing this method
+    # to blow up and cause the server to continually fail to start.
+    # This is called very early from evm_application and ensures we attempt to call MiqWorker#kill
+    # to remove the problematic rows.
+    handle_orphaned_worker_classes do
```

Comment:
Alternatively, we could detect a `ActiveRecord::SubclassNotFound` and try grouping the worker rows by the type columns and loop over each type and only "fix" the type column that it's failing on.  The code here was already too much and I was afraid grouping by type would be even worse.  😢

Reply:
What do you think?

Reply:
I think way you have it now is simpler than grouping by worker, though perhaps it should be more isolated to just the query instead of the query and the kill block (then perhaps you don't need a separate method)?  Something like...

```ruby
workers =
  begin
    miq_workers.where(:status => MiqWorker::STATUSES_CURRENT_OR_STARTING).to_a
  rescue ActiveRecord::SubclassNotFound
    MiqWorker.inheritance_column = nil
    retry
  ensure
    MiqWorker.inheritance_column = "type"
  end

workers.each do |w|
  w.kill
  ...
end
```

Reply:
Now that I wrote that I'm concerned there's a bug.  Since the query is for *all* workers, then if even 1 worker is invalid, then all workers get queried without STI.  So, what happens if there is a custom kill method?

Reply:
Exactly, that's why I was pondering "trying" the happy path and if any of the workers fail to instantiate with their subclass, we can retry but grouped by "type" and if one type fails, we'll only retry without STI for that class.

This feels needlessly complicated since `kill_all_workers` the custom kill method for subclasses do things like add pid file deletion to the super method, which shouldn't really be a big deal.  🤷‍♂ 

I can go either way.

Reply:
We decided to merge our thoughts, and build up a list of bad worker types by collecting all the types of known workers and find any that aren't in that list... then use the disable sti trick to use the base class to kill them, then we can through the remaining workers and call the normal kill on them.